man/io_uring_prep_accept.3

```diff
@@ -103,26 +90,68 @@ and see if
 .B IORING_CQE_F_MORE
 is set on completion as an indication of whether or not the accept request
 will generate further CQEs. Note that for the multishot variants, setting
-.B addr
+.I addr
 and
-.B addrlen
+.I addrlen
 may not make a lot of sense, as the same value would be used for every
 accepted connection. This means that the data written to
-.B addr
+.I addr
 may be overwritten by a new connection before the application has had time
 to process a past connection. If the application knows that a new connection
 cannot come in before a previous one has been processed, it may be used as
 expected. The multishot variants are available since 5.19.
 
-For multishot with direct descriptors,
-.B IORING_FILE_INDEX_ALLOC
-must be used as the file descriptor. This tells io_uring to allocate a free
-direct descriptor from our table, rather than the application passing one in.
-Failure to do so will result in the accept request being terminated with
-.BR -EINVAL .
-The allocated descriptor will be returned in the CQE
+If a direct variant is used, the application must first have registered
+a file table using
+.BR io_uring_register_files (3)
+of the appropriate size. Once registered,
+.BR io_uring_prep_accept_direct (3)
+may specify any
+.I file_index
+of that table, as long as it is within the size of the registered table.
+If the specified entry already contains a file, the file will first be removed
+from the table and closed, consistent with the behavior of updating an
+existing file with
+.BR io_uring_register_files_update (3).
+
+.B TODO
```

Comment:
If there's no specific opcode to test for and no feature flag for it, the best solution is probably as you suggest and check for IORING_OP_SOCKET support. That isn't perfect as you could have an older kernel that backported either one of them, and you'd either get a false negative or false positive by checking for socket support. But probably good enough for most cases.

Reply:
I'm reworking my contributions to this page now. I don't like my last version.

Reply:
Am reading through other man pages today to see how they have dealt with the direct helpers, and I realize I shouldn't be diverging very far from the original where it used the same copied descriptions of the direct feature. There could be some other attempt at scrubbing all the pages and getting them all to refer to a common direct file table man page some day. Unlike when I was totally new to urings and was going through the man pages, I think the man pages would be better in the long run if they didn't repeat themselves so often. The important aspects of each function or function helper is lost when the majority of the description is mostly repeated from other pages. And as features like multi and direct evolve, it would be nice for a single page to reflect the latest description. But for today, let me get a new draft out of this one page.

Reply:
I agree, longer term it would be nice to have topic man pages for things like fixed files, registered buffers, provided buffers, etc. And also agree on the duplicated content. A lot of that would be solved with the topic man pages for areas of interest.

Reply:
Will you create a link for io_uring_register_files_sparse.3 so I can refer to it on this page? I'm torn between not mentioning it because other pages don't either or getting it right on at least this page for now.

Reply:
I'm confusing myself with ideas on how the application can safeguard itself if it is running on older kernels. The original warning is about older kernels not checking the *file_index* field but shouldn't it be enough to have first checked that a direct table could be registered with one of the two library functions for registering a table?

Reply:
Should the date of the manpage be changed or does that indicate when the feature was first introduced?

Reply:
You can just create the link as a separate commit.

Reply:
> but shouldn't it be enough to have first checked that a direct table could be registered with one of the two library functions for registering a table?

Registered files were available since the initial release, so way before eg the requests that can instantiate one.